.sandbox/provision/lifecycle-asg/on-delete.sh

```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+set -ex
+
+source ./common.sh
+
+INSTANCE_ID="$(get_instance_id)"
+VOLUME_ID="$(get_volume_id)"
+
+if [ -n "$INSTANCE_ID" ] ; then
+    aws ec2 terminate-instances --instance-ids "$INSTANCE_ID"
+    aws ec2 wait instance-terminated --instance-ids "$INSTANCE_ID"
+fi
+
+if [ -n "$VOLUME_ID" ]; then
+    aws ec2 delete-volume --volume-id "$VOLUME_ID"
+fi
+
+./cleanup.sh
```

Comment:
This script may not work well with retry-on-failure: if L11,12 failed once, this script will always fail (e.g. the instance has been already deleted) and leaving the volume leaked. Note: sandbox resource has a limited number of retry of deletion.

A better solution will be ignoring the error of `aws ec2 terminate-instances` and `aws ec2 wait instance-terminated`, and always try `aws ec2 delete-volume`. Let it fail if delete volume failed for another retry, so the volume will be eventually deleted with certain number of retries.

Reply:
If the instance has been already deleted, `get_instance_id` should not retrieve a valid value and L11-12 should be skipped. The check of volume and instance are separated so volume should not be leaked

Reply:
If the resource leaking is caused by the limited number of retries, the suggested solution would not address it neither. Since the delete-volume command would fail 3 time in the row. The best option is to increase the retry count or introducing some final resource cleaning mechanism.

Reply:
I don't fully agree on this. The script should try best effort to reduce the friction. The mostly likely case causing this failure will be the EC2 instance was deleted out-of-band before running the delete hook. In this case, if the script can move on and delete the volume, it will handle over 90% cases perfectly. Otherwise, the user will need to handle this all the time when happened. The impression of the product is fragile.